tree/ntuple/v7/src/RFieldMeta.cxx

```diff
@@ -197,6 +198,37 @@ void ROOT::Experimental::RClassField::ReadInClusterImpl(RClusterIndex clusterInd
    }
 }
 
+void ROOT::Experimental::RClassField::BeforeConnectPageSource(Internal::RPageSource &pageSource)
+{
+   if (GetOnDiskId() == kInvalidDescriptorId) {
+      return;
+   }
+   // Gather all known sub fields in the descriptor.
+   std::unordered_set<std::string> knownSubFields;
+   {
+      const auto descriptorGuard = pageSource.GetSharedDescriptorGuard();
+      const RNTupleDescriptor &desc = descriptorGuard.GetRef();
+      const auto &fieldDesc = desc.GetFieldDescriptor(GetOnDiskId());
+      // Check that we have the same type; only then we perform automatic schema evolution.
+      // TODO: should this throw? Right now, we allow reading back a field hierarchy with a different type as long as
+      // all columns match...
+      if (GetTypeName() != fieldDesc.GetTypeName())
+         return;
+
+      for (auto linkId : fieldDesc.GetLinkIds()) {
+         const auto &subFieldDesc = desc.GetFieldDescriptor(linkId);
+         knownSubFields.insert(subFieldDesc.GetFieldName());
+      }
+   }
+
+   // Iterate over all sub fields in memory and mark those as missing that are not in the descriptor.
+   for (auto &field : fSubFields) {
+      if (knownSubFields.count(field->GetFieldName()) == 0) {
+         field->SetArtificial();
+      }
```

Comment:
Base classes we should match by type name, rather than by field name.

Reply:
This ties to the TODO above: for the moment we allow reading back
 * models with matching field names and generated columns, in principle with different type names, and
 * models with additional fields as long as the surrounding type name matches (because only then we mark it as artificial).

So for example, a field of type `Derived` in
```c++
struct BaseA {
  int fInt;
};
struct Derived : public BaseA {};
```
would automatically evolve to
```c++
struct BaseB {
  int fInt;
};
struct Derived : public BaseB {};
```
which I think is as expected. On the other hand, it would also transparently evolve into
```c++
struct BaseA {
  int fInt;
};
struct BaseB {
  int fInt;
};
struct Derived : public BaseB, public BaseA {};
```
where `BaseA::fInt` is read into `BaseB::fInt` because `BaseB` comes first...

Reply:
As discussed offline, we generally forbid the case of mismatching type names of classes for the moment.